File: 310-minimum-height-trees/minimum-height-trees.py
class Solution(object):
    def findMinHeightTrees(self, n, edges):
        if n == 1 or not edges:
            return [0]
        
        res = []
        x = defaultdict(set)

        for u, v in edges:
            x[u].add(v)
            x[v].add(u)
        
        for i, j in x.items():
            if len(j) == 1:
                res.append(i)
        
        while n > 2:
            n -= len(res)
            y = []
            for k in res:
                u = next(iter(x[k]))
                x[u].remove(k)
                if len(x[u]) == 1:
                    y.append(u)
            res = y
        
        return res
        

File: 310-minimum-height-trees/minimum-height-trees.java
import java.util.*;

class Solution {
    public List<Integer> findMinHeightTrees(int n, int[][] edges) {
        List<Integer> res = new ArrayList<>();
        if (n == 1) {
            res.add(0);
            return res;
        }
        
        if (edges.length == 0) {
            for (int i = 0; i < n; i++) {
                res.add(i);
            }
            return res;
        }
        
        Map<Integer, Set<Integer>> x = new HashMap<>();
        for (int[] edge : edges) {
            int u = edge[0];
            int v = edge[1];
            if (!x.containsKey(u)) {
                x.put(u, new HashSet<>());
            }
            x.get(u).add(v);
            if (!x.containsKey(v)) {
                x.put(v, new HashSet<>());
            }
            x.get(v).add(u);
        }
        
        for (int i : x.keySet()) {
            if (x.get(i).size() == 1) {
                res.add(i);
            }
        }
        
        while (n > 2) {
            n -= res.size();
            List<Integer> y = new ArrayList<>();
            for (int k : res) {
                int u = x.get(k).iterator().next();
                x.get(u).remove(k);
                if (x.get(u).size() == 1) {
                    y.add(u);
                }
            }
            res = y;
        }
        
        return res;
    }
}

